kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,794 @@
+local DEFAULT_STALE_TTL   = 4
+local DEFAULT_EMPTY_TTL   = 30
+-- long-lasting TTL of 10 years for hosts or static IP addresses in cache settings
+local LONG_LASTING_TTL    = 10 * 365 * 24 * 60 * 60
+
+local DEFAULT_ORDER = { "LAST", "SRV", "A", "AAAA", "CNAME" }
+
+local TYPE_SRV      = resolver.TYPE_SRV
+local TYPE_A        = resolver.TYPE_A
+local TYPE_AAAA     = resolver.TYPE_AAAA
+local TYPE_CNAME    = resolver.TYPE_CNAME
+local TYPE_LAST     = -1
+
+local NAME_TO_TYPE = {
+  SRV     = TYPE_SRV,
+  A       = TYPE_A,
+  AAAA    = TYPE_AAAA,
+  CNAME   = TYPE_CNAME,
+  LAST    = TYPE_LAST,
+}
+
+local TYPE_TO_NAME = {
+  [TYPE_SRV]      = "SRV",
+  [TYPE_A]        = "A",
+  [TYPE_AAAA]     = "AAAA",
+  [TYPE_CNAME]    = "CNAME",
+}
+
+local HIT_L3 = 3 -- L1 lru, L2 shm, L3 callback, L4 stale
+
+local HIT_LEVEL_TO_NAME = {
+  [1] = "hit_lru",
+  [2] = "hit_shm",
+  [3] = "hit_cb",
+  [4] = "hit_stale",
+}
+
+-- server replied error from the DNS protocol
+local NAME_ERROR_CODE             = 3 -- response code 3 as "Name Error" or "NXDOMAIN"
+-- client specific error
+local CACHE_ONLY_ERROR_CODE       = 100
+local CACHE_ONLY_ERROR_MESSAGE    = "cache only lookup failed"
+local CACHE_ONLY_ANSWERS = tablex.readonly({ errcode = CACHE_ONLY_ERROR_CODE, errstr = CACHE_ONLY_ERROR_MESSAGE })
+local EMPTY_RECORD_ERROR_CODE     = 101
+local EMPTY_RECORD_ERROR_MESSAGE  = "empty record received"
+
+
+-- APIs
+
+local _M = {}
+local mt = { __index = _M }
+
+-- copy TYPE_*
+for k,v in pairs(resolver) do
+  if type(k) == "string" and k:sub(1,5) == "TYPE_" then
+    _M[k] = v
+  end
+end
+_M.TYPE_LAST = TYPE_LAST
+
+
+local TRIES_MT = { __tostring = cjson.encode, }
+
+
+local function stats_init(stats, name)
+  if not stats[name] then
+    stats[name] = {}
+  end
+end
+
+
+local function stats_count(stats, name, key)
+  stats[name][key] = (stats[name][key] or 0) + 1
+end
+
+
+local function stats_set(stats, name, key, value)
+  stats[name][key] = value
+end
+
+
+-- lookup or set TYPE_LAST (the DNS record type from the last successful query)
+local function insert_last_type(cache, name, qtype)
+  local key = "last:" .. name
+  if TYPE_TO_NAME[qtype] and cache:get(key) ~= qtype then
+    cache:set(key, { ttl = 0 }, qtype)
+  end
+end
+
+
+local function get_last_type(cache, name)
+  return cache:get("last:" .. name)
+end
+
+
+-- insert hosts into cache
+local function init_hosts(cache, path, preferred_ip_type)
+  local hosts, err = parse_hosts(path)
+  if not hosts then
+    log(WARN, "Invalid hosts file: ", err)
+    hosts = {}
+  end
+
+  if not hosts.localhost then
+    hosts.localhost = {
+      ipv4 = "127.0.0.1",
+      ipv6 = "[::1]",
+    }
+  end
+
+  local function insert_answer(name, qtype, address)
+    if not address then
+      return
+    end
+
+    -- insert via the `:get` callback to prevent inter-process communication
+    cache:get(name .. ":" .. qtype, nil, function()
+      return {
+        ttl = LONG_LASTING_TTL,
```

Comment:
@bungle With your previous comment about why it introduced `ttl` field into answers, I just use to temporary value that told mlcache how long to save it. So I think I could pass it to upper caller as a return value while not as a field `answers.ttl`, I will update it later.

Reply:
After some effort, I failed to remove `ttl` field from `answers`, because we could not get `ttl` value from the return value of `mlcache:get()` in `resolve_name_type`, it can only return one value like `answers`, so we could not insert this `answers` into mlcache as a short key.

The original DNS library could do this because it will re-calculate the item ttl before inserting it to the cache. So it will caculate ttl twice for inserting it as key and short-key.

Because we have utilized mlcache to make the layers of get-key logic and get-short-key logic clearer, I think it's better if we calculate the TTL just after querying it from the DNS server. So here I keep the `answers.ttl` field.